Maple2.Server.Game/Util/DamageCalculator.cs

```diff
@@ -0,0 +1,146 @@
+﻿using Maple2.Model.Enum;
+using Maple2.Model.Metadata;
+using Maple2.Server.Core.Formulas;
+using Maple2.Server.Game.Model;
+using Maple2.Server.Game.Model.Skill;
+using Maple2.Server.Game.Packets;
+
+namespace Maple2.Server.Game.Util;
+
+public static class DamageCalculator {
+    public static (DamageType, long) CalculateDamage(IActor caster, IActor target, DamageRecord? record = null, bool dotDamage = false) {
+        // Check block
+        if (Damage.TriggerCompulsionEvent(target.Buffs.TotalCompulsionRate(CompulsionEventType.BlockChance, record?.SkillId ?? 0))) {
+            return (DamageType.Block, 0);
+        }
+
+        // Check evase
+        if (Damage.TriggerCompulsionEvent(target.Buffs.TotalCompulsionRate(CompulsionEventType.EvasionChanceOverride, record?.SkillId ?? 0))) {
+            return (DamageType.Miss, 0);
+        }
+
+        // Get hit rate
+        if (Damage.Evade(target.Buffs.GetResistance(BasicAttribute.Accuracy), target.Stats.Values[BasicAttribute.Evasion].Total, target.Stats.Values[BasicAttribute.Dexterity].Total,
+                caster.Stats.Values[BasicAttribute.Accuracy].Total, caster.Buffs.GetResistance(BasicAttribute.Evasion), caster.Stats.Values[BasicAttribute.Luck].Total)) {
+            return (DamageType.Miss, 0);
+        }
+
+        var damageType = DamageType.Normal;
+        (double minBonusAtkDamage, double maxBonusAtkDamage) = caster.Stats.GetBonusAttack(target.Buffs.GetResistance(BasicAttribute.BonusAtk), target.Buffs.GetResistance(BasicAttribute.MaxWeaponAtk));
+        double attackDamage = minBonusAtkDamage + (maxBonusAtkDamage - minBonusAtkDamage) * Random.Shared.NextDouble();
```

Comment:
I feel like instead if minBonus and maxBonus it should just be minAttack and maxAttack. Bonus damage is its own stat with its own implications, while the output of this function is just broader results of all attack calculations. Also might need a way to disable the randomness cause `isConstDotDamageValue` & `isConstDamageValue` both exist. I have not looked into exactly how these work, but they're something to keep in mind.

Reply:
I believe we are using `isConstDotDamageValue`. The damage calculation will never happen and would just use the HpValue defined. is that correct?

Reply:
Oh I'm actually not sure. Sounds right enough